client/src/http_default.rs

```diff
@@ -134,46 +108,81 @@ impl RequestBuilder for DefaultRequestBuilder {
         K: AsRef<str>,
         V: ToString,
     {
-        Self {
-            params: Some(
-                params
-                    .into_iter()
-                    .map(|pair| {
-                        let (k, v) = pair.borrow();
-                        (k.as_ref().to_owned(), v.to_string())
-                    })
-                    .collect(),
-            ),
-            ..self
-        }
+        self.and_then(|b| Ok(b.params(params)))
+    }
+
+    fn body(self, data: Option<Vec<u8>>) -> Self::Output {
+        self.0.map(|b| {
+            DefaultRequest::new(b.bytes(match data {
+                Some(bytes) => bytes,
+                None => Vec::new(),
+            }))
+        })
     }
 }
 
-pub type WebSocketStream = WebSocket<MaybeTlsStream<TcpStream>>;
+/// Special builder for WS connections
+pub struct DefaultWebSocketRequestBuilder(Result<http::request::Builder>);
+
+impl DefaultWebSocketRequestBuilder {
+    fn and_then<F>(self, func: F) -> Self
```

Comment:
Would consider making this a trait, so that opaque representations can still be chained.